splink/input_column.py

```diff
@@ -1,112 +1,207 @@
+        is_string = not is_castable_to_int(index_name)
+        literal = exp.Literal(this=index_name, is_string=is_string)
+
+        self.bracket_index = exp.Bracket(expressions=[literal])
+        return self
 
+    def append_expression(self, expression: exp.Expression):
+        self.column_expressions.append(expression)
+        self.remove_duplicate_expressions()
+        return self
+
+    def remove_duplicate_expressions(self):
+        """
+        Remove duplicate expressions from the column expressions list.
 
-def remove_quotes_from_identifiers(tree) -> Expression:
-    tree = tree.copy()
-    for identifier in tree.find_all(exp.Identifier):
-        identifier.args["quoted"] = False
-    return tree
+        Ordering is determined by the sequence of appends; the most recently added
+        items to the list are retained and appear as the outermost branches
+        in the SQL expression tree.
+        """
+        seen_types = set()
+        # Iterate from right to left
+        for index in range(len(self.column_expressions) - 1, -1, -1):
+            if type(self.column_expressions[index]) in seen_types:
+                del self.column_expressions[index]
+            else:
+                seen_types.add(type(self.column_expressions[index]))
+
+        return self.column_expressions
+
+    @property
+    def column_tree(self):
+        # Create our column representation
+        column = sqlglot.column(col=self.name, table=self.table, quoted=self.quoted)
+        # Column is made up of the identifier + bracket index (if it exists)
+        if self.bracket_index:
+            self.bracket_index.set("this", column)
+            return self.bracket_index
+        else:
+            return column
+
+    def build_column_tree_with_expressions(self) -> sqlglot.Expression:
+        """Constructs an SQLglot expression tree by combining a column expression
+        with a sequence of SQLglot expressions.
+
+        This method iteratively sets the column value for each expression,
+        starting from the provided column and extending to the most recently
+        added expression.
+
+        Expressions are modified in-place to avoid copying. This only works
+        for expressions expecting a single child - i.e. any value that only
+        expects a single "this" key.
+        """
+        column = self.column_tree
+        column_expressions = deepcopy(self.column_expressions)
+
+        if column_expressions and self.return_column_with_expressions:
+            # Recursively select the innermost element of the expression tree and
+            # add it to our sqlglot column through a simple find and replace.
+
+            # This is possible as our expressions are all functions expecting a single
+            # column argument, which we can add a placeholder for and replace.
+            for expression in column_expressions:
+                expression.find(exp.Column).replace(column)
```

Comment:
I like this idea, but I'm not sure it works as soon as we have non-sqlglot compatible expressions e.g. if we've registered our own custom functions.  Maybe it does?  

I wonder whether we get more flexibility if we use sql strings here rather than sqlglot expression trees.